packages/flutter/lib/src/rendering/object.dart

```diff
@@ -3248,31 +3260,64 @@ abstract class RenderObject with DiagnosticableTreeMixin implements HitTestTarge
   /// the global coordinate system in logical pixels. To get physical pixels,
   /// use [applyPaintTransform] from the [RenderView] to further transform the
   /// coordinate.
-  Matrix4 getTransformTo(RenderObject? ancestor) {
-    final bool ancestorSpecified = ancestor != null;
+  Matrix4 getTransformTo(RenderObject? target, { bool skipIfInUnpaintedSubtree = false }) {
     assert(attached);
-    if (ancestor == null) {
-      final RenderObject? rootNode = owner!.rootNode;
-      if (rootNode is RenderObject) {
-        ancestor = rootNode;
+    // The paths to fromRenderObject and toRenderObject's common ancestor.
+    List<RenderObject>? fromPath;
+    List<RenderObject>? toPath;
+
+    RenderObject from = this;
+    RenderObject to = target ?? owner!.rootNode!;
+
+    while (!identical(from, to)) {
+      final int fromDepth = from.depth;
+      final int toDepth = to.depth;
+
+      if (fromDepth >= toDepth) {
+        assert(from.parent != null, '$target and $this are not in the same render tree.');
+        final RenderObject fromParent = from.parent!;
+        if (skipIfInUnpaintedSubtree && !fromParent.paintsChild(from)) {
+          return Matrix4.zero();
+        }
+        (fromPath ??= <RenderObject>[this]).add(fromParent);
+        from = fromParent;
+      }
+      if (fromDepth <= toDepth) {
+        assert(to.parent != null, '$target and $this are not in the same render tree.');
+        assert(target != null);
+        final RenderObject toParent = to.parent!;
+        if (skipIfInUnpaintedSubtree && !toParent.paintsChild(to)) {
+          return Matrix4.zero();
+        }
+        (toPath ??= target == null ? <RenderObject>[] : <RenderObject>[target]).add(toParent);
+        to = toParent;
       }
     }
-    final List<RenderObject> renderers = <RenderObject>[];
-    for (RenderObject renderer = this; renderer != ancestor; renderer = renderer.parent!) {
-      renderers.add(renderer);
-      assert(renderer.parent != null); // Failed to find ancestor in parent chain.
+
+    Matrix4? fromTransform;
+    if (fromPath != null) {
+      assert(fromPath.length > 1);
+      fromTransform = Matrix4.identity();
+      final int lastIndex = target == null ? fromPath.length - 2 : fromPath.length - 1;
+      for (int index = lastIndex; index > 0; index -= 1) {
+        fromPath[index].applyPaintTransform(fromPath[index - 1], fromTransform);
+      }
     }
-    if (ancestorSpecified) {
-      renderers.add(ancestor!);
+    if (toPath == null) {
+      return fromTransform ?? Matrix4.identity();
     }
-    final Matrix4 transform = Matrix4.identity();
-    for (int index = renderers.length - 1; index > 0; index -= 1) {
-      renderers[index].applyPaintTransform(renderers[index - 1], transform);
+
+    assert(toPath.length > 1);
```

Comment:
If `target == null` couldn't the length of the toPath just be 1? what garantees that the while loop above executes at least twice to guarantee `> 1`?

Reply:
each list must have more than one element if not null.